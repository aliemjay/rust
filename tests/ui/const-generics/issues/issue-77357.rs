#![feature(generic_const_exprs)]
#![allow(incomplete_features)]

trait MyTrait<T> {}

fn bug<'a, T>() -> &'static dyn MyTrait<[(); { |x: &'a u32| { x }; 4 }]> {
    //~^ ERROR overly complex generic constant
    //~| ERROR cycle detected when evaluating type-level constant
    todo!()
}

fn main() {}
