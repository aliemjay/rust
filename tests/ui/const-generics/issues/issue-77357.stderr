error: overly complex generic constant
  --> $DIR/issue-77357.rs:6:46
   |
LL | fn bug<'a, T>() -> &'static dyn MyTrait<[(); { |x: &'a u32| { x }; 4 }]> {
   |                                              ^^^^^^^^^^^^^^^^^^^^^^^^^ blocks are not supported in generic constants
   |
   = help: consider moving this anonymous constant into a `const` function
   = note: this operation may be supported in the future

error[E0391]: cycle detected when evaluating type-level constant
  --> $DIR/issue-77357.rs:6:46
   |
LL | fn bug<'a, T>() -> &'static dyn MyTrait<[(); { |x: &'a u32| { x }; 4 }]> {
   |                                              ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: ...which requires const-evaluating + checking `bug::{constant#0}`...
  --> $DIR/issue-77357.rs:6:46
   |
LL | fn bug<'a, T>() -> &'static dyn MyTrait<[(); { |x: &'a u32| { x }; 4 }]> {
   |                                              ^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires const-evaluating + checking `bug::{constant#0}`...
  --> $DIR/issue-77357.rs:6:46
   |
LL | fn bug<'a, T>() -> &'static dyn MyTrait<[(); { |x: &'a u32| { x }; 4 }]> {
   |                                              ^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires caching mir of `bug::{constant#0}` for CTFE...
  --> $DIR/issue-77357.rs:6:46
   |
LL | fn bug<'a, T>() -> &'static dyn MyTrait<[(); { |x: &'a u32| { x }; 4 }]> {
   |                                              ^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires elaborating drops for `bug::{constant#0}`...
  --> $DIR/issue-77357.rs:6:46
   |
LL | fn bug<'a, T>() -> &'static dyn MyTrait<[(); { |x: &'a u32| { x }; 4 }]> {
   |                                              ^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires borrow-checking `bug::{constant#0}`...
  --> $DIR/issue-77357.rs:6:46
   |
LL | fn bug<'a, T>() -> &'static dyn MyTrait<[(); { |x: &'a u32| { x }; 4 }]> {
   |                                              ^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: ...which requires normalizing `Binder(ConstEvaluatable(Const { ty: usize, kind: Unevaluated(UnevaluatedConst { def: WithOptConstParam { did: DefId(0:8 ~ issue_77357[fee6]::bug::{constant#0}), const_param_did: None }, substs: [T] }) }), [])`...
   = note: ...which again requires evaluating type-level constant, completing the cycle
note: cycle used when computing the implied bounds of `bug`
  --> $DIR/issue-77357.rs:6:1
   |
LL | fn bug<'a, T>() -> &'static dyn MyTrait<[(); { |x: &'a u32| { x }; 4 }]> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0391`.
